File: src/Pcl.NET/PCDReader.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Pcl.NET
{
  public class PCDReader : UnmanagedObject
  {
    public PCDReader()
    {
      _ptr = Invoke.io_pcdreader_ctor();
    }

    public void Read(string fileName, PointCloudXYZ cloud, int offset = 0)
    {
      int err = Invoke.io_pcdreader_read_xyz(_ptr, fileName, cloud.Ptr, offset);
#warning TODO check err
    }

    protected override void DisposeObject()
    {
      Invoke.io_pcdreader_delete(ref _ptr);
    }
  }
}

File: src/pcl-bindings/PCDReader.cpp
#include "PCDReader.h"

PCDReader* io_pcdreader_ctor()
{
    return new PCDReader();
}

void io_pcdreader_delete(PCDReader** ptr)
{
    delete* ptr;
    *ptr = NULL;
}

int32_t io_pcdreader_read_xyz(PCDReader* ptr, const char* str, PointCloud<PointXYZ>* cloud, int offset)
{
    return ptr->read(string(str), *cloud, offset);
}
